README.md
# Fan control balenaBlock

![](./logo.png)
 
This block can be used to controls a cooling fan based on a system thermal zones (temperature measurement managed on Linux systems).

The block supports 2 types of fan control strategies:
- GPIO based ON/OFF control, with an hysteresis,
- PWM based control, with speed slopes.

**NOTE: As linux kernel manages temperatures in Celsius, all temperatures in this block configuration are in Celsius (and integer value).**

## Compatibility

The primary role for this block is to allow for CPU fan control by user-space software, which is easier to re-configure on the fly than kernel based fan control, and thus more suitable for hacking projects.

As long as your senors appear in `/sys/class/thermal/` as a `thermalzone`, and the GPIO/PWM you want to use is not in use by other configuration, this block should be compatible.

## Configuration

The block is configured through environment variables passed to the service. Available variables are:
- For **GPIO mode**:
  - `FAN_GPIO`: The GPIO number your fan is connected to, see your specific board reference to get this number (default value `132`),
  - `THERMAL_HIGH`: The "high" temperature, the fan will turn ON when temperature reaches this level (default value `60`),
  - `THERMAL_LOW`: The "low" temperature, the fan will turn OFF when temperature reaches this level (default value `40`). 

- For **PWM mode**:
  - `PWM`: a comma-separated address of the `PWMCHIP/PWM` to use. _e.g._ `1,0` will use `/sys/class/pwm/pwmchip1/pwm` to control the PWM (no default value, setting this value disables **GPIO mode** and enables **PWM mode**),
  - `PWM_PERIOD`: period to configure your `pwmX/period` to, this will be the maximum value of your duty cycle. (default is `40000`),
  - `PWM_TEMPS`: a comma-separated list of configuration temperature points.These points are used for linear interpolation for the PWM control curve (default `0,40,50,70`).
  - `PWM_DUTY`: a comma-separated list of duty-cycles to apply for each of previously defined temperature points. **This list should be the same length as `PWM_TEMPS`, with each value in the range 0-`PWM_PERIOD`.** (default `0,200,10000,40000`).

- For `both modes`:
  - `REFRESH_DELAY`: the period, in seconds, between two fan speed update (default value `60`),
  - `THERMAL_ZONE`: the thermal zone device number on your system you want to use as a source for fan control (default value `0`).

### Additional configuration

You also will need to set your service as `privileged` and enable `io.balena.features.sysfs=1` in your labels.

## Example configuration in GPIO mode

```yaml
services:
  fan-control:
      image: bh.cr/balena/fan_control_arm
      privileged: true
      restart: always
      labels:
         io.balena.features.sysfs: 1
      environment:
         - 'REFRESH_DELAY=30'
         - 'GPIO="14"'
         - 'THERMAL_ZONE=0'
         - 'THERMAL_LOW=45'
    		 - 'THERMAL_HIGH=65'
```

## Example configuration in PWM mode

```yaml
services:
  fan-control:
      image: bh.cr/balena/fan_control_arm
      privileged: true
      restart: always
      labels:
         io.balena.features.sysfs: 1
      environment:
         - 'THERMAL_ZONE=1'
         - 'REFRESH_DELAY=10'
         - 'PWM="1,0"'
         - 'PWM_TEMPS="0,40,50,70"'
         - 'PWM_DUTY="0,100,1000,40000"'
```




CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.1.1
## (2022-11-11)

* Adding restrart policy mechanism [Aurelien VALADE]

# v0.1.0
## (2022-11-09)

* Adding repo.yml for versionist management [Aurelien VALADE]
* Adding balena_slugs to the project in flowzone.yml [Aurelien VALADE]
* Cleanning up documentation for first release [Aurelien VALADE]


fan-control.sh
#!/bin/bash


# Exits on failure
set -e

# Get environment variables, set default values to match etcherPro
FAN_GPIO=${FAN_GPIO:-132}
THERMAL_ZONE=/sys/class/thermal/thermal_zone${THERMAL_ZONE:-0}/temp
THERMAL_HIGH=${THERMAL_HIGH:-60}000
THERMAL_LOW=${THERMAL_LOW:-40}000
REFRESH_DELAY=${REFRESH_DELAY:-60}
PWM_PERIOD=${PWM_PERIOD:-40000}
PWM_TEMPS=${PWM_TEMPS:-"0,40,50,70"}
PWM_DUTY=${PWM_DUTY:-"0,200,10000,40000"}

setup_gpio() {
# First, set GPIO as output
if [[ -d /sys/class/gpio/gpio${FAN_GPIO} ]]; then
   echo -n ${FAN_GPIO} > /sys/class/gpio/unexport
fi

echo -n ${FAN_GPIO} > /sys/class/gpio/export
echo -n out > /sys/class/gpio/gpio${FAN_GPIO}/direction
echo -n 1 > /sys/class/gpio/gpio${FAN_GPIO}/value
}

setup_pwm() {
  IFS=','
  read PWM_CHIP PWM_NUM <<< ${PWM}
  if ! [[ -d /sys/class/pwm/pwmchip${PWM_CHIP}/pwm${PWM_NUM} ]]; then
	  echo -n ${PWM_NUM} > /sys/class/pwm/pwmchip${PWM_CHIP}/export
  fi

  
  read -ra TEMPS <<< ${PWM_TEMPS}
  read -ra DUTY <<< ${PWM_DUTY}
  export TEMPS
  export DUTY
  export PWM_PATH=/sys/class/pwm/pwmchip${PWM_CHIP}/pwm${PWM_NUM}

  if [[ $(cat ${PWM_PATH}/enable) -eq 1 ]];then
	  echo -n 0 > ${PWM_PATH}/enable
  fi
  echo -n ${PWM_PERIOD} > ${PWM_PATH}/period
  echo -n ${DUTY[0]} > ${PWM_PATH}/duty_cycle
  echo -n 1 > ${PWM_PATH}/enable
}

if [[ -z ${PWM} ]]; then
    echo "Setting up GPIO mode"
    setup_gpio
else
    echo "Setting up PWM mode"
	setup_pwm
fi

regulate_gpio() {
	TEMP=$1

	if [[ ${TEMP} -ge ${THERMAL_HIGH} ]];then
       echo -n 1 > /sys/class/gpio/gpio${FAN_GPIO}/value
	elif [[ ${TEMP} -le ${THERMAL_LOW} ]];then
       echo -n 0 > /sys/class/gpio/gpio${FAN_GPIO}/value
	fi
}

regulate_pwm() {
    TEMP=$1
    
	# First, look for saturations
	if [[ ${TEMP} -le ${TEMPS[0]}000 ]];then
		echo -n ${DUTY[0]} > ${PWM_PATH}/duty_cycle
		return
	elif [[ ${TEMP} -ge ${TEMPS[-1]}000 ]];then
		echo -n ${DUTY[-1]} > ${PWM_PATH}/duty_cycle
		return
	fi

	# Otherwise, regulate
	for t in ${!TEMPS[@]}; do
		# First, we check the where the temperature sits in our list
		if [[ ${TEMP} -le ${TEMPS[$t]}000 ]]; then
			# When we found the segment ($t (T2) is the index of the temperature above the measurement in the table
			# $t-1 (T1) is the temperature below)
			# We add 3 tailing zeros to be compatible with kernel temperature format.
			T1=${TEMPS[$(($t - 1))]}000
			T2=${TEMPS[$t]}000
			DC1=${DUTY[$(($t - 1))]}
			DC2=${DUTY[$t]}
			# We want to do a linear interpolation between the corresponding PWM duty cycles.
			# - First, we get the position ratio (between 0 and 1). As BASH arithmetic is integer only, it will be between 0 and 1000
			#   R = (TEMP - T1)/(T2 - T1)
			RATIO=$(( ($TEMP - $T1) * 1000 /($T2 - $T1) ))
			# - Then we interpolate the duty cycle value 
			DC=$(($DC1 + ($DC2 - $DC1) * $RATIO / 1000 ))

			if ! [[ -z ${DEBUG} ]];then
			   echo "Temp ${TEMP}, Ratio ${RATIO}, Duty ${DC}"
			fi
			echo -n ${DC} > ${PWM_PATH}/duty_cycle
			return
		fi
	done
}

while :; do
	TEMP=$(cat ${THERMAL_ZONE}) || exit 1
    if [[ -z ${PWM} ]];then
		regulate_gpio $TEMP
	else
		regulate_pwm $TEMP
	fi
	sleep ${REFRESH_DELAY}
done
